src/Domain/Services/ExtendRestApi.php

```diff
@@ -29,14 +38,26 @@ class ExtendRestApi {
 	/**
 	 * Constructor
 	 *
+	 * @param Container  $container An instance of the container class.
 	 * @param Package    $package An instance of the package class.
 	 * @param Formatters $formatters An instance of the formatters class.
 	 */
-	public function __construct( Package $package, Formatters $formatters ) {
+	public function __construct( Container $container, Package $package, Formatters $formatters ) {
```

Comment:
I don't think we should pass in the container. It works, but the proper pattern (I think) would be to pass in a schema repository class that is used to retrieve the schema. It reduces the scope of making the entire container available to this class. 

But I'm not sure how the schema repository class would then be created/implemented. It's kind of a chicken/egg scenario at risk of creating circular dependencies because ideally `SchemaController` would be passed into `ExtendRestApi` but we have `ExtendRestAPI` passed into the `SchemaController` AND each of the schemas 🤔 .

Until we can come up with a better solution, an interim approach for now (even though it's also an antipattern - service locator) would be to just call `Package::container()` directly in the new `getSchema` method here. And we can add a todo to improve this. That way we're not creating a signature for the constructor that we might change significantly down the road.

**The larger question here**, is what need is there for a `get_schema` method on this class (I know it's currently implemented in the companion Subscriptions extension pr)? Shouldn't any code needing schemas use the `SchemaController` directly? I'm pretty inclined to revisit this approach as I'm pretty wary about the cyclical dependency this creates.

Reply:
Speaking of which, I wonder if we should make this class final?

Reply:
I did it this way because of the reason you mentioned; circular dependency. Subs needs access to our schema definitions via extend, but the schema definitions themselves are passed the extend class. I wasn't sure if there was another way around this.

Reply:
Ya I get that. Still to avoid creating a signature that's hard to walk back from, I think for now it'd be wise to do a couple things here:

- Make the `ExtendRestAPI` class final.
- Directly get the container from `Package:container` for now in `get_schema` and add a todo as tech debt to revisit in the future. 

Thoughts?

Reply:
Actually I think we should maybe think about how, or if, Extend should provide Schema to extensions, because if we include `get_schema ` for extensions we cannot back out of that contract.

@senadir did you have any thoughts on this?

I wonder if we could rejig how extend works so rather than it being passed to schema classes, it's invoked when schema classes are used by the routes.

Reply:
> Actually I think we should maybe think about how, or if, Extend should provide Schema to extensions, because if we include get_schema for extensions we cannot back out of that contract.

I believe we discussed this before; what value do we gain by exposing our schema to extensions? We're neither validating the developer schema nor responsible for rendering it.

One argument for this is that having the same schema would make code reuse much easier but would put a higher maintenance cost for us, giving that we're committing to our REST API shape (and therefore schema), this should be fine.

> I wonder if we could rejig how extend works so rather than it being passed to schema classes, it's invoked when schema classes are used by the routes.

I don't get this part? This is how it works right now, extend hook into `get_properties` and `get_item_response` which are evaluated when the route actually use them.

Reply:
@senadir Right now extend works within each schema `get_item_response ` - I was suggesting considering moving the extend logic to the Route where `get_item_response` is called.

Reply:
So something like this?
```php
return rest_ensure_response( array_merge( $this->schema->get_item_response( $cart ), $this->extend->get_extended_schema( self::IDENTIFIER ) ) );
```

Reply:
@senadir I'd add a utility to merge, but yes, that. Routes are not something we'll pass to extensions, but schema we may, so this avoids circular dependencies  _because schema will have no dependencies._

Reply:
@senadir Not as straightforward as hoped as some schemas use their own schema to derive props. Lets leave as is—I'll look for alternative solution.

Reply:
I've reverted the change and used it directly in subs https://github.com/woocommerce/woocommerce-subscriptions/pull/3959/commits/ff158451e38e997dfd7bbc10020428940ba6a080